Add omf commands for listing available plugins/themes and for downloading a plugin/theme

farao opened the pull request:
It would be nice to have a possibility to download themes and plugins from the oh-my-fish repository and to list all available ones.

This is a proposal how it could be done. It adds the following commands to omf:
* available-plugins (shows all plugins from github repository)
* available-themes (shows all themes from github repository)
* download-plugin PLUGIN (uses omf.packages.install to download and install a plugin from the github repo)
* download-theme THEME (uses omf.packages.install to download and install a theme from the github repo)

This is not only for human use: e.g. the new "omf.packages --download-theme" could be used by the "theme" plugin to automatically download a theme when it's not locally available (without duplicating code but with the nice "Installing..." message).

I'm not fixed on the naming or on how things work and I'd be happy about feedback. Also I'm not a fish expert, so I'm not quite sure if the coding style meets the usual conventions.

bpinto commented:
@farao This is super!

I'd like to discuss the CLI though. What do you think about:

> available-plugins (shows all plugins from github repository)

`omf plugins list` or just `omf plugins`

> available-themes (shows all themes from github repository)

`omf themes list` or just `omf themes`

> download-plugin PLUGIN (uses omf.packages.install to download and install a plugin from the github repo)

`omf install` already does this.

> download-theme THEME (uses omf.packages.install to download and install a theme from the github 
repo)

`omf install` already does this.

bpinto commented on plugins/omf/omf.packages.fish:
```diff
@@ -51,11 +59,33 @@ function omf.packages --argument-names options -d 'Manage all plugins and themes
   end
 end
 
+function get_all_repos
```
This could become a separate function called `omf.packages.remote` which receives `--plugins` or `--themes` as arguments and return the list names in an array.

farao commented:
Yeah probably "omf plugins" and "omf themes" is the best way? I can change that!

Regarding download-plugin and download-theme: If I understood it right, "omf install" takes the list of plugins and the current theme from the config and downloads them if necessary. But there is no way to do "omf install theme-willy" or so, right? Without changing the config. Without using it. Just downloading it for local use. Espescially for the theme plugin it would be nice to say: "this plugin is not available locally, shall I download it?"! Or is that also possible via omf install?

bpinto commented:
You are right. I misread it.

Perhaps `omf install $name`?

For keeping the task simple, I don't think we have a case where a plugin name is the same as a theme name. But in case you want to work on it, I would maybe add support for `omf install --plugin $name`?

farao commented:
Ok, I moved the get_all_repos function to an own file as omf.remote. Also I changed the cli:
* omf install -> like before
* omf install --theme|--plugin NAME -> installs a theme or plugin
* omf plugins -> shows all available plugins
* omf themes -> shows all available themes

bucaran commented:
@bpinto Good call on trying to simplify things :+1: Still feels a bit too complex, but it's looking better now.

farao commented:
the omf install with arguments could maybe be left out from the main cli and if the functionality is needed, you can still use omf.packages --install --theme NAME ? I guess it's more an internal functionality anyways once theme can download themes automatically?!

bpinto commented:
I wanted to remove the `theme` plugin and bring it to `omf`.

So maybe `omf theme $name` to download/set it as current theme.
And `omf theme --default $name` to download/set it as current theme and also change the `config.fish` file.

bpinto commented on plugins/omf/omf.packages.fish:
```diff
@@ -51,6 +64,10 @@ function omf.packages --argument-names options -d 'Manage all plugins and themes
         omf.log normal $fish_theme
       end
+    case '--plugins'
+      omf.remote --plugins | paste -sd ' ' -
```
To keep it dry isn't it better to move `| paste -sd ' ' -` to `omf.remote`?

farao commented:
Oh, yes, that was duplicated anyways. Just forgot to delete it, did that now

farao commented:
Is there already functionality to write a theme into the config automatically?

bpinto commented:
Nope, I think this PR solves one problem. Listing remote themes and plugins. Want to merge this and work on the next one?

If you come online on gitter, it might be easier!

farao commented:
Sure, might be better in a new PR!

bpinto commented:
Apparently `sed -r` is not valid option for sed on osx. I have tried the commands locally and they fail. =/

bpinto commented:
Things left:

- [ ] Remove option to install a theme/plugin from the CLI help interface.
- [ ] Replace GNU specific sed for a generic one `grep "^Link" | sed 's/Link:.*page=\([0-9]*\)&per_page=100>; rel="last".*/\1/')`

farao commented:
Should be everything...

bpinto commented:
Thanks @farao
